### DISCUSSION TITLE
Firmware rollback / Where to get EU_V3.1.0_BF2-0

### DISCUSSION THREAD

**Discussion by Ohrenmann** [toniebox-reverse-engineering/hackiebox_cfw_ng#30] [Q&A] [answered]
Hi,
I'd like to use some of the patches. According to the documentation: 
_"currently the latest version supported for all patches is EU_V3.1.0_BF2-0"_

Unfortunately my TonieBox already received a newer firmware version ota before I started the experiments.
Is it possible to get the original firmware legally somewhere and perform a firmware downgrade?

If not, how can I learn to create patches for newer firmware on my own?

BR

**Comment by SciLor** [comment, marked as answer]
Hey there,
you usually have 3 different firmware versions on your box. Please make a full dump and check it with the [firmware_info Tool](https://github.com/toniebox-reverse-engineering/toniebox/blob/master/tools/firmware_info.py).
Which firmware version do you have?

The latest known version is EU_V3.1.0_BF4, but it is defective, and the box always reverts to EU_V3.1.0_BF2.

To make your own patches you'll have to reverse engineer the firmware with Ghidra for example and then find the right spots to change the firmwares' behavior.

**Reply by Ohrenmann** [reply]
Many thanks for the answer. I already forgot that I did the backups. 
First was EU_V3.1.0_BF4
Second and third were are version EU_V3.1.0_BF2.
Copied the second one from backup to sd:/revvox/boot/ng-ofw1.bin and sd:/revvox/boot/ng-ofw2.bin and edited the config to launch ng-ofw2.bin as default.

In theory that's it.
